collector/net.go

```diff
@@ -0,0 +1,178 @@
+	nicBlacklist = flag.String("collector.network.nic-blacklist", "", "Regexp of NIC:s to blacklist. NIC name must both match whitelist and not match blacklist to be included.")
+)
+
+// A NetworkCollector is a Prometheus collector for WMI Win32_PerfRawData_Tcpip_NetworkInterface metrics
+type NetworkCollector struct {
+	// XXX names
+	BytesReceivedPerSec *prometheus.Desc
+	BytesSentPerSec     *prometheus.Desc
+	BytesTotalPerSec    *prometheus.Desc
+
+	CurrentBandwidth *prometheus.Desc
+
+	PacketsPerSec         *prometheus.Desc
+	PacketsReceivedPerSec *prometheus.Desc
+	PacketsSentPerSec     *prometheus.Desc
+
+	nicWhitelistPattern *regexp.Regexp
+	nicBlacklistPattern *regexp.Regexp
+}
+
+// NewNetworkCollector ...
+func NewNetworkCollector() (Collector, error) {
+	const subsystem = "net"
+
+	return &NetworkCollector{
+		BytesReceivedPerSec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "bytes_received_per_sec"),
+			"(Network.BytesReceivedPerSec)",
+			[]string{"nic"},
+			nil,
+		),
+		BytesSentPerSec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "bytes_sent_per_sec"),
+			"(Network.BytesSentPerSec)",
+			[]string{"nic"},
+			nil,
+		),
+
+		BytesTotalPerSec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "bytes_total_per_sec"),
+			"(Network.BytesTotalPerSec)",
+			[]string{"nic"},
+			nil,
+		),
+		CurrentBandwidth: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "current_bandwidth"),
+			"(Network.CurrentBandwidth)",
+			[]string{"nic"},
+			nil,
+		),
+		PacketsPerSec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "packets_per_sec"),
+			"(Network.PacketsPerSec)",
+			[]string{"nic"},
+			nil,
+		),
+		PacketsReceivedPerSec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "packets_received_per_sec"),
+			"(Network.PacketsReceivedPerSec)",
+			[]string{"nic"},
+			nil,
+		),
+		PacketsSentPerSec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "packets_sent_per_sec"),
+			"(Network.PacketsSentPerSec)",
+			[]string{"nic"},
+			nil,
+		),
+
+		nicWhitelistPattern: regexp.MustCompile(fmt.Sprintf("^(?:%s)$", *nicWhitelist)),
+		nicBlacklistPattern: regexp.MustCompile(fmt.Sprintf("^(?:%s)$", *nicBlacklist)),
+	}, nil
+}
+
+// Collect sends the metric values for each metric
+// to the provided prometheus Metric channel.
+func (c *NetworkCollector) Collect(ch chan<- prometheus.Metric) error {
+	if desc, err := c.collect(ch); err != nil {
+		log.Println("[ERROR] failed collecting iis metrics:", desc, err)
+		return err
+	}
+	return nil
+}
+
+type Win32_PerfRawData_Tcpip_NetworkInterface struct {
+	CurrentBandwidth      uint32
+	BytesReceivedPerSec   uint32
+	BytesSentPerSec       uint32
+	BytesTotalPerSec      uint64
+	Name                  string
+	PacketsPerSec         uint32
+	PacketsReceivedPerSec uint32
+	PacketsSentPerSec     uint32
+}
```

Comment:
Some suggested additions, since they seem to provide a bit of error detection:
```
PacketsOutboundDiscarded
PacketsOutboundErrors
PacketsReceivedDiscarded
PacketsReceivedErrors
PacketsReceivedUnknown (?)
```